lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,638 @@
+                f"Dataset state {self.state} is terminal and cannot be advanced to {new_state}"
+            )
+        elif new_state not in self.transitions[self.state]:
+            self.logger.error(
+                "Current state {} can't advance to {}", self.state, new_state
+            )
+            raise DatasetBadStateTransition(
+                f"Dataset desired state {new_state} is not allowed from current state {self.state}"
+            )
+
+        # TODO: this would be a good place to generate an audit log
+
+        self.state = new_state
+        self.transition = datetime.datetime.now()
+        self.update()
+
+    def add(self):
+        """
+        add Add the Dataset object to the database
+        """
+        try:
+            Database.db_session.add(self)
+            Database.db_session.commit()
+        except Exception as e:
+            self.logger.error("Can't add {} to DB", str(self))
+            Database.db_session.rollback()
+            raise DatasetSqlError(f"Error adding {self}: {e}")
+
+    def update(self):
+        """
+        update Update the database row with the modified version of the
+        Dataset object.
+        """
+        try:
+            Database.db_session.commit()
+        except Exception as e:
+            self.logger.error("Can't update {} in DB", str(self))
+            Database.db_session.rollback()
+            raise DatasetSqlError(f"Error updating {self}: {e}")
+
+
+class Metadata(Database.Base):
+    """ Retain secondary information about datasets
+
+    Columns:
+        id          Generated unique ID of table row
+        dataset_ref Dataset row ID (foreign key)
+        key         Metadata key string
+        value       Metadata value string
+    """
+
+    __tablename__ = "metadata"
+
+    # Currently defined metadata keys
+    REINDEX = "REINDEX"
+    ARCHIVED = "ARCHIVED"
+    TARBALL_PATH = "TARBALL_PATH"
+
+    METADATA_KEYS = [REINDEX, ARCHIVED, TARBALL_PATH]
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    key = Column(String(255), unique=False, nullable=False, index=True)
+    value = Column(String(2048), unique=False, nullable=True)
+    dataset_ref = Column(
+        Integer, ForeignKey("dataset.id", ondelete="CASCADE"), nullable=False
+    )
+
+    def __init__(self, **kwargs):
+        """
+        Construct a new Metadata object for a dataset
+
+        Parameters:
+
+            key     Metadata key (from METADATA_KEYS)
+            value   String value for key
+        """
+        try:
+            key = kwargs.get("key")
+        except KeyError:
+            raise MetadataBadKey("Metadata key is required")
+        else:
+            if key not in Metadata.METADATA_KEYS:
+                raise MetadataBadKey(f"Metadata key {key} is not supported")
+        try:
+            value = kwargs.get("value")
+        except KeyError:
+            raise MetadataBadKey(f"Metadata key {key} value is required")
```

Comment:
(LOL, I'm a little obtuse...or I'm reviewing too much code....)

`dict:get()` [never raises an exception](https://python-reference.readthedocs.io/en/latest/docs/dict/get.html), so the `try` blocks are pointless and wrong.  :-D

So, the code should be something more like this:
```
        key = kwargs.get("key")
        if key is None:
            raise MetadataBadKey("Metadata key is required")
        if key not in Metadata.METADATA_KEYS:
            raise MetadataBadKey(f"Metadata key {key} is not supported")
        value = kwargs.get("value")
        if value is None: 
            raise MetadataBadKey(f"Metadata key {key} value is required")
```
No `try` blocks (or nasty `else`'s) required!  :-)

Reply:
Ha; I'd originally used subscripted access; and when I changed to `get` I didn't even think to fix the surrounding code to adapt! Good catch.